### PR TITLE
Address 3160, 3161 and 3162

### PR DESCRIPTION
**Pull request by davidkline-ms**
- Fixes: #3160
- Fixes: #3161 
- Fixes: #3162 (also investigating a better, longer term fix)

### CODE REVIEW

**Comment by davidkline-ms** [comment]
These are literal string compares. It feels as if the tests (as modified) accomplish the same goal (check the return value / object state) without relying upon specific text strings being logged. It feels like one of them would have failed even before the change (was expecting a word not in the actual log message).

I would be open for an issue to track this.

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs [hunk 1]
```diff
@@ -549,13 +584,19 @@ public bool RegisterService(Type type, IMixedRealityService service)
 
             if (type == null)
             {
-                DebugUtilities.LogError("Unable to add a manager of type null.");
+                if ((ActiveProfile.LoggingLevel & LoggingLevels.Error) != 0)
+                {
+                    Debug.LogError("Unable to add a manager of type null.");
```
[the remark this answers is not in the document]
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]: Can this be a separate PR? These were the strings that previously existed.

File: Assets/MixedRealityToolkit-Tests/MixedRealityToolkitTests.cs [hunk 2]
```diff
@@ -178,7 +176,6 @@ public void Test11_TestMixedRealityExtensionService2DoesNotReturn()
 
             // Validate non-existent component
             MixedRealityToolkit.Instance.GetService(typeof(ITestComponent2), "Test2");
-            LogAssert.Expect(LogType.Error, "Unable to find Test2 Manager.");
```
[the remark this answers is not in the document]
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit-Tests/MixedRealityToolkitTests.cs]: The test should probably not be checking the log string then (requires an exact match), and checking the return from GetService.

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]: No, let's get it taken care of. We're already modifying the lines anyway.

File: Assets/MixedRealityToolkit-Tests/MixedRealityToolkitTests.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit-Tests/MixedRealityToolkitTests.cs]: That's how Unity's test suite works. 

Please fix the test.

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs [hunk 3]
```diff
@@ -376,8 +395,12 @@ private void InitializeInstance()
                 {
                     if (instance != null)
                     {
-                        DebugUtilities.Assert(instance.transform.parent == null, "The MixedRealityToolkit should not be parented under any other GameObject!");
-                        DebugUtilities.Assert(instance.transform.childCount == 0, "The MixedRealityToolkit should not have GameObject children!");
+                        if ((ActiveProfile.LoggingLevel & LoggingLevels.Assert) != 0)
+                        {
+
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]: Please remove extra line

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs [hunk 4]
```diff
@@ -549,13 +584,19 @@ public bool RegisterService(Type type, IMixedRealityService service)
...
+                }
                 return false;
             }
 
             if (service == null)
             {
-                DebugUtilities.LogError("Unable to add a manager with a null instance.");
+                if ((ActiveProfile.LoggingLevel & LoggingLevels.Error) != 0)
+                {
+                    Debug.LogError("Unable to add a manager with a null instance.");
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]:
```suggestion
                    Debug.LogError("Unable to add a service with a null instance.");
```

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs [hunk 5]
```diff
@@ -616,13 +663,19 @@ public IMixedRealityService GetService(Type type, bool showLogs = true)
 
             if (!IsInitialized)
             {
-                DebugUtilities.LogCriticalError($"Unable to get {nameof(type)} Manager as the Mixed Reality Manager has not been initialized!");
+                if ((ActiveProfile.LoggingLevel & LoggingLevels.CriticalError) != 0)
+                {
+                    Debug.LogError($"Critical: Unable to get {nameof(type)} Manager as the Mixed Reality Manager has not been initialized!");
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]:
```suggestion
                    Debug.LogError($"Critical: Unable to get {nameof(type)} service as the Mixed Reality Toolkit has not been initialized!");
```

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs [hunk 6]
```diff
@@ -616,13 +663,19 @@ public IMixedRealityService GetService(Type type, bool showLogs = true)
...
+                }
                 return null;
             }
 
             if (type == null)
             {
-                DebugUtilities.LogError("Unable to get null manager type.");
+                if ((ActiveProfile.LoggingLevel & LoggingLevels.Error) != 0)
+                {
+                    Debug.LogError("Unable to get null manager type.");
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]:
```suggestion
                    Debug.LogError("Unable to get null service type.");
```

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs [hunk 7]
```diff
@@ -660,13 +716,19 @@ public IMixedRealityService GetService(Type type, string serviceName, bool showL
...
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]:
```suggestion
                    Debug.LogError("Unable to get null service type.");
```

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs [hunk 8]
```diff
@@ -660,13 +716,19 @@ public IMixedRealityService GetService(Type type, string serviceName, bool showL
...
+                }
                 return null;
             }
 
             if (string.IsNullOrEmpty(serviceName))
             {
-                DebugUtilities.LogError("Unable to get manager by name without the name being specified.");
+                if ((ActiveProfile.LoggingLevel & LoggingLevels.Error) != 0)
+                {
+                    Debug.LogError("Unable to get manager by name without the name being specified.");
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]:
```suggestion
                    Debug.LogError("Unable to get service by name without the name being specified.");
```

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs [hunk 9]
```diff
@@ -699,7 +764,10 @@ public void UnregisterService(Type type)
 
             if (type == null)
             {
-                DebugUtilities.LogCriticalError("Unable to remove null manager type.");
+                if ((ActiveProfile.LoggingLevel & LoggingLevels.Error) != 0)
+                {
+                    Debug.LogError("Unable to remove null manager type.");
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]:
```suggestion
                    Debug.LogError("Unable to remove null service type.");
```

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs [hunk 10]
```diff
@@ -683,7 +745,10 @@ public IMixedRealityService GetService(Type type, string serviceName, bool showL
 
             if (service == null && showLogs)
             {
-                DebugUtilities.LogWarning($"Unable to find {serviceName} Manager.");
+                if ((ActiveProfile.LoggingLevel & LoggingLevels.Warning) != 0)
+                {
+                    Debug.LogWarning($"Unable to find {serviceName} service.");
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]: > Let's do in a separate PR

But we already made a [partial change](https://github.com/Microsoft/MixedRealityToolkit-Unity/pull/3164/files#diff-f21536a45f2f15aa366accfb4e1e7d9bL686). Let's go all the way.

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs [hunk 11]
```diff
@@ -731,13 +799,19 @@ public void UnregisterService(Type type, string serviceName)
 
             if (type == null)
             {
-                DebugUtilities.LogCriticalError("Unable to remove null manager type.");
+                if ((ActiveProfile.LoggingLevel & LoggingLevels.Error) != 0)
+                {
+                    Debug.LogError("Unable to remove null manager type.");
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]:
```suggestion
                    Debug.LogError("Unable to remove null service type.");
```

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs [hunk 12]
```diff
@@ -731,13 +799,19 @@ public void UnregisterService(Type type, string serviceName)
...
+                }
                 return;
             }
 
             if (string.IsNullOrEmpty(serviceName))
             {
-                DebugUtilities.LogCriticalError("Unable to remove manager by name without the name being specified.");
+                if ((ActiveProfile.LoggingLevel & LoggingLevels.Error) != 0)
+                {
+                    Debug.LogError("Unable to remove manager by name without the name being specified.");
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]:
```suggestion
                    Debug.LogError("Unable to remove service by name without the name being specified.");
```

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs [hunk 13]
```diff
@@ -764,7 +838,10 @@ public void DisableService(Type type)
         {
             if (type == null)
             {
-                DebugUtilities.LogError("Unable to disable null manager type.");
+                if ((ActiveProfile.LoggingLevel & LoggingLevels.Error) != 0)
+                {
+                    Debug.LogError("Unable to disable null manager type.");
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]:
```suggestion
                    Debug.LogError("Unable to disable null service type.");
```

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs [hunk 14]
```diff
@@ -792,13 +869,19 @@ public void DisableService(Type type, string serviceName)
...
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]:
```suggestion
                    Debug.LogError("Unable to disable null service type.");
```

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs [hunk 15]
```diff
@@ -792,13 +869,19 @@ public void DisableService(Type type, string serviceName)
...
+                }
                 return;
             }
 
             if (string.IsNullOrEmpty(serviceName))
             {
-                DebugUtilities.LogError("Unable to disable manager by name without the name being specified.");
+                if ((ActiveProfile.LoggingLevel & LoggingLevels.Error) != 0)
+                {
+                    Debug.LogError("Unable to disable manager by name without the name being specified.");
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]:
```suggestion
                    Debug.LogError("Unable to disable service by name without the name being specified.");
```

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs [hunk 16]
```diff
@@ -825,7 +908,10 @@ public void EnableService(Type type)
         {
             if (type == null)
             {
-                DebugUtilities.LogError("Unable to enable null manager type.");
+                if ((ActiveProfile.LoggingLevel & LoggingLevels.Error) != 0)
+                {
+                    Debug.LogError("Unable to enable null manager type.");
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]:
```suggestion
                    Debug.LogError("Unable to enable null service type.");
```

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs [hunk 17]
```diff
@@ -853,13 +939,19 @@ public void EnableService(Type type, string serviceName)
...
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]:
```suggestion
                    Debug.LogError("Unable to enable null service type.");
```

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs [hunk 18]
```diff
@@ -853,13 +939,19 @@ public void EnableService(Type type, string serviceName)
...
+                }
                 return;
             }
 
             if (string.IsNullOrEmpty(serviceName))
             {
-                DebugUtilities.LogError("Unable to enable manager by name without the name being specified.");
+                if ((ActiveProfile.LoggingLevel & LoggingLevels.Error) != 0)
+                {
+                    Debug.LogError("Unable to enable manager by name without the name being specified.");
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]:
```suggestion
                    Debug.LogError("Unable to enable service by name without the name being specified.");
```

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs [hunk 19]
```diff
@@ -891,7 +983,10 @@ public List<IMixedRealityService> GetActiveServices(Type type)
         {
             if (type == null)
             {
-                DebugUtilities.LogWarning("Unable to get managers with a type of null.");
+                if ((ActiveProfile.LoggingLevel & LoggingLevels.Error) != 0)
+                {
+                    Debug.LogError("Unable to get managers with a type of null.");
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]:
```suggestion
                    Debug.LogError("Unable to get service with a type of null.");
```

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs [hunk 20]
```diff
@@ -910,7 +1005,10 @@ public List<IMixedRealityService> GetActiveServices(Type type, string serviceNam
 
             if (type == null)
             {
-                DebugUtilities.LogWarning("Unable to get managers with a type of null.");
+                if ((ActiveProfile.LoggingLevel & LoggingLevels.Error) != 0)
+                {
+                    Debug.LogError("Unable to get managers with a type of null.");
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]:
```suggestion
                    Debug.LogError("Unable to get service with a type of null.");
```

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs [hunk 21]
```diff
@@ -1075,7 +1173,10 @@ private static bool IsCoreSystem(Type type)
         {
             if (type == null)
             {
-                DebugUtilities.LogWarning("Null cannot be a core manager.");
+                if ((Instance?.ActiveProfile.LoggingLevel & LoggingLevels.Error) != 0)
+                {
+                    Debug.LogError("Null cannot be a core manager.");
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]:
```suggestion
                    Debug.LogError("Null cannot be a core system!");
```

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs [hunk 22]
```diff
@@ -1104,7 +1205,10 @@ private void GetService(Type type, out IMixedRealityService service)
         {
             if (type == null)
             {
-                DebugUtilities.LogWarning("Unable to get a component with a type of null.");
+                if ((ActiveProfile.LoggingLevel & LoggingLevels.Error) != 0)
+                {
+                    Debug.LogError("Unable to get a component with a type of null.");
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]:
```suggestion
                    Debug.LogError("Unable to get a service with a type of null.");
```

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs [hunk 23]
```diff
@@ -1122,7 +1226,10 @@ private bool GetService(Type type, string serviceName, out IMixedRealityService
...
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]:
```suggestion
                    Debug.LogError("Unable to get a service with a type of null.");
```

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs [hunk 24]
```diff
@@ -1156,7 +1266,10 @@ private void GetService(Type type, ref List<IMixedRealityService> services)
         {
             if (type == null)
             {
-                DebugUtilities.LogWarning("Unable to get components with a type of null.");
+                if ((ActiveProfile.LoggingLevel & LoggingLevels.Error) != 0)
+                {
+                    Debug.LogError("Unable to get components with a type of null.");
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]:
```suggestion
                    Debug.LogError("Unable to get services with a type of null.");
```

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs [hunk 25]
```diff
@@ -1167,7 +1280,10 @@ private void GetServices(Type type, string serviceName, ref List<IMixedRealitySe
...
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]:
```suggestion
                    Debug.LogError("Unable to get services with a type of null.");
```

File: Assets/MixedRealityToolkit-Tests/MixedRealityToolkitTests.cs [hunk 26]
```diff
@@ -33,8 +33,6 @@ public void Test01_InitializeMixedRealityToolkit()
         [Test]
         public void Test02_TestNoMixedRealityConfigurationFound()
         {
-            LogAssert.Expect(LogType.Error, "No Mixed Reality Configuration Profile found, cannot initialize the Mixed Reality Toolkit");
```
[the remark this answers is not in the document]
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit-Tests/MixedRealityToolkitTests.cs]: This doesn't break the test case.. The Assert.IsFalse(MixedRealityToolkit.HasActiveProfile);  and Assert.IsNull(MixedRealityToolkit.Instance.ActiveProfile);  are the true tests and are less fragile (a one character change in the code breaks string based comparisons like this).

File: Assets/MixedRealityToolkit-Tests/MixedRealityToolkitTests.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit-Tests/MixedRealityToolkitTests.cs]: I will implement a less fragile version of the test by checking that the value returned by GetService is null, using Assert.IsTrue()

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs [hunk 27]
```diff
@@ -177,7 +178,10 @@ private void InitializeServiceLocator()
                 if (!RegisterService(typeof(IMixedRealityInputSystem), Activator.CreateInstance(ActiveProfile.InputSystemType) as IMixedRealityInputSystem) ||
                     InputSystem == null)
                 {
-                    DebugUtilities.LogCriticalError("Failed to start the Input System!");
+                    if ((ActiveProfile.LoggingLevel & LoggingLevels.CriticalError) != 0)
```
[the remark this answers is not in the document]
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]: Yeah it really seems like we're just duplicating the existing functionality of the editor...

https://docs.unity3d.com/ScriptReference/Logger.html

File: Assets/MixedRealityToolkit-Tests/MixedRealityToolkitTests.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit-Tests/MixedRealityToolkitTests.cs]: > We also test the errors we "Expect" when something is not correct. 

While I agree that this is a good endeavor, it can make the tests fragile. If a method changes the log string (ex: a spelling error is fixed, the strings get localized, or if we add a feature like this one, that allows developers to disable some log messages) the test will fail.

One possible solution is to return error _codes_ and validate those values, rather than a specific log string.

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]: Fair enough (and Stephen already changed the values).

File: Assets/MixedRealityToolkit-Tests/MixedRealityToolkitTests.cs (hunk 26 above)
[the remark this answers is not in the document]
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit-Tests/MixedRealityToolkitTests.cs]: Part of the feedback on this PR was to consolidate the message that occurs when ActiveProfile is null (have the getter log the messsage). That resulted in the system displaying an error when the MixedRealityToolkit component is first added to the scene (by default, there is no profile configured).

Is the preference to go back to logging a message in each caller if they detect a null ActiveProfile? I am fine with that.

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs [hunk 28]
```diff
@@ -69,17 +69,18 @@ public MixedRealityToolkitConfigurationProfile ActiveProfile
...
-                }
 #if UNITY_EDITOR
                 if (!Application.isPlaying && activeProfile == null)
                 {
                     UnityEditor.Selection.activeObject = Instance;
                     UnityEditor.EditorGUIUtility.PingObject(Instance);
                 }
+#else
+                // If there is no active profile, the system cannot function. _Always_ display this message in the debug output.
+                if (activeProfile == null)
```
[the remark this answers is not in the document]
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]: > this call should NOT have been moved from it's original location,

That change was made based on feedback... I can easily re-add the log messages at the points of access.

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs (hunk 27 above)
[the remark this answers is not in the document]
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]: We received customer feedback to enable a path to configure the volume of the logging from the toolkit. This PR's pattern has been successfully leveraged in past projects.

> why would Anyone ever want to "hide" critical errors?

Adding the ability to filter critical errors is about customer choice. Some developers do not wish to see the error logs and simply handle return values to check for errors.

[the remark this answers is not in the document]
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]: I completely disagree. The fact that there's a critical error needs to be addressed by the developer in question or the whole system will not work. If they don't wish the critical error in question, about the status of the _entire_ system, then they can prevent this message by removing the toolkit all-together.

[the remark this answers is not in the document]
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]: TL;DR Errors like these shouldn't be simply ignored, but fixed and addressed.

**Comment by davidkline-ms** [comment]
Looking at this:

109158-unity-logger-update

It appears that unity provides a similar solution (with a few more clicks). I'll approve the rollback PR (#3167) and check back with the customer.
